src/expiring_lru.rs

```diff
@@ -64,39 +45,95 @@ where
     type Target = V;
 
     fn deref(&self) -> &Self::Target {
-        match &self.0 {
-            EntryRefInner::RefMut(inner) => &inner.value().inner,
-            EntryRefInner::RefMulti(inner) => &inner.value().inner,
-        }
+        &self.0.value().inner
     }
 }
 
-async fn expire_entries<K, V>(map: ExpiringLru<K, V>, reap_interval: Duration, expiration: Duration)
-where
-    K: Eq + Hash,
+async fn decay_task<K, V>(
+    map: ExpiringLru<K, V>,
+    expiration: Duration,
+    mut rx: UnboundedReceiver<TimerUpdate<K>>,
+) where
+    K: Eq + Hash + Clone + Send + Sync + 'static,
+    V: Send + Sync + 'static,
 {
-    let mut interval = interval(reap_interval);
+    let mut queue = DelayQueue::new();
 
     loop {
-        interval.tick().await;
-
-        let right_now = Instant::now();
-        map.inner
-            .retain(|_, entry| entry.last_used + expiration > right_now);
-
-        debug!("Done reaping timed out HTTP ratelimiters");
+        tokio::select! {
+            expired = queue.next(), if !queue.is_empty() => {
+                // An item expired in the queue, remove it from the map
+                if let Some(key) = expired {
+                    debug!("Removing expired entry from ratelimiter decay queue");
+                    map.remove(key.get_ref());
+                } else {
+                    // This should not occur because we only poll next if the queue is not empty
+                    break;
+                }
+            }
+            msg = rx.recv() => {
+                if let Some(msg) = msg {
+                    match msg {
+                        TimerUpdate::Add { map_key, return_key_to } => {
+                            debug!("Adding entry to ratelimiter decay queue");
+                            let key = queue.insert(map_key, expiration);
+                            let _ = return_key_to.send(key);
+                        },
+                        TimerUpdate::Refresh { key } => {
+                            debug!("Refreshing entry in ratelimiter decay queue");
+                            // This will panic if the key is not present, therefore
+                            // we check that in the calling end
+                            queue.reset(&key, expiration);
+                        },
+                        TimerUpdate::Remove { key } => {
+                            debug!("Removing entry in ratelimiter decay queue");
+                            queue.try_remove(&key);
+                        }
+                        TimerUpdate::RemoveLru { return_map_key_to } => {
+                            debug!("Removing least recently used item from ratelimiter decay queue");
+                            if let Some(expired) = queue.peek().and_then(|key| queue.try_remove(&key)) {
```

Comment:
here we call `queue.try_remove`, so why does this return the key just so that it sender can issue a `TimerUpdate::Remove` request (which calls `queue.try_remove` again). There seems to be no need to return the key nor for `ExpiringLru::remove_lru` to call `remove`

Reply:
There actually is a need to return the key. This part only removes  the entry from the queue, not from the map. Therefore `ExpiringLru::remove_lru` has to remove it from the map, but you are right that it can avoid sending a `TimerUpdate::Remove`.